src/hammer-vlsi/defaults.yml

```diff
@@ -231,31 +231,45 @@ vlsi.inputs:
   # Each item is a struct member:
   # path (str) - Path to the given instance
-  # type (str) - One of the following:
-  # - "dummy" (does nothing with this constraint)
-  # - "placement" (creates a placement constraint for an instance)
-  # - "toplevel" (top-level chip dimensions; may only occur once, for the top-level module)
-  # - "hardmacro" (places this hard macro at a particular spot)
-  # - "hierarchical" (marks this instance as part of hierarchical place and route)
-  # - "obstruction" (creates a blockage at a particular spot; see obs_types)
-  # TODO(edwardw): add rotation
+  # - Required for all types
+  # type (str) - The type of placement constraint
+  # - Required for all types
+  # - One of the following:
+  #   - "dummy" (does nothing with this constraint)
+  #   - "placement" (creates a placement constraint for an instance)
+  #   - "toplevel" (top-level chip dimensions; may only occur once, for the top-level module)
+  #   - "hardmacro" (places this hard macro at a particular spot)
+  #   - "hierarchical" (marks this instance as part of hierarchical place and route)
+  #   - "obstruction" (creates a blockage at a particular spot; see obs_types)
+  # orientation (Optional[str]) - The orientation (MX, MY, R0, R90, R180, R270)
+  # - Optional for all types
   # x (float) - x coordinate in um
+  # - Required for all types
   # y (float) - y coordinate in um
+  # - Required for all types
   # width (float) - width in um
+  # - Required for all types, but can be auto-filled for hierarchical and hardmacro if left blank
   # height (float) - height in um
-  # margins (Optional[Margins]) - margins for the top-level module (i.e. only only for a toplevel constraint). Contains the following keys:
-  # - "left" - margin from the left side of the core PNR area to the edge of the chip
-  # - "right" - margin from the right side of the core PNR area to the edge of the chip
-  # - "top" - margin from the top side of the core PNR area to the edge of the chip
-  # - "bottom" - margin from the bottom side of the core PNR area to the edge of the chip
-  # top_layer (str) - Optional: Specifies the highest layer used by this hardmacro
-  #   If specified will be used to keep other wires away from this hardmacro
-  # layers (List[str]) - Optional: only used for obstruction type. 
-  #   If specified, a list of strings that enumerates layer(s) blocked by the obstruction otherwise all layers are blocked
-  # obs_types (List[str]) - Required for obstruction type. A list of one or more of the following:
-  # - "place" - This obstruction type stops the placement of standard cells inside it
-  # - "route" - This obstruction type stops all routing inside it
-  # - "power" - This obstruction type stops only power routing/straps inside it
+  # - Required for all types, but can be auto-filled for hierarchical and hardmacro if left blank
+  # master (Optional[str]) - The master module name of the hierarchical or hardmacro constraint
```

Comment:
So is the idea something like this?

Somewhere in some master list (it's not clear to me how this list ends up in Hammer yet), there's a list "my_macro -> (400, 100)".
Then we create a hardmacro instantation like
```
path: "MyTop/amplifier_inst"
type: hardmacro
x: 100
y: 100
master: my_macro
```

I think I get what's going on, would be good to elaborate on this in the documentation and add a unit test.

Reply:
`master` is a pretty common name for the module an instance is an instance of; I don't really think it needs to be elaborated. I can go into more detail about where the info comes from, though. For `hierarchical` it comes from the `toplevel` placement constraint of that block, and for `hardmacro` it comes from `vlsi.technology.extra_macro_sizes` (hence the discussion yesterday).

Reply:
(Forgot to add that I will put this in the documentation)

Reply:
Maybe if you can just put the example in the documentation that may be good enough. I had to do a double take to really understand what this example is doing, and I'd like for Hammer to be as easy to use as possible. If it's too much work for you I can help

Reply:
Added some documentation, please take a look. Working on adding a unit test.